Lib/asyncio/executor.py

```diff
@@ -0,0 +1,226 @@
+
+        end_time = None if timeout is None else time.monotonic() + timeout
+
+        inputs_stream = _azip(*iterables)
+        submitted_tasks = Queue[Optional[Future]]()
+        tasks_in_flight_limit = len(self._workers) + self._input_queue.maxsize
+        resume_feeding = Event()
+
+        async def _feeder() -> None:
+            try:
+                async for args in inputs_stream:
+                    if self._shutdown:
+                        break
+                    future = await self.submit(fn, *args)
+                    await submitted_tasks.put(future)
+
+                    if submitted_tasks.qsize() >= tasks_in_flight_limit:
+                        await resume_feeding.wait()
+                    resume_feeding.clear()
+            except QueueShutDown:
+                # The executor was shut down while feeder waited to submit a
+                # task.
+                pass
+            finally:
+                await submitted_tasks.put(None)
+                submitted_tasks.shutdown()
+
+        feeder_task = create_task(_feeder())
+        self._feeders.add(feeder_task)
+        feeder_task.add_done_callback(self._feeders.remove)
+
+        try:
+            while True:
+                task = await submitted_tasks.get()
+                if task is None:
+                    break
+
+                remaining_time = (
+                    None if end_time is None else end_time - time.monotonic()
+                )
+                if remaining_time is not None and remaining_time <= 0:
+                    raise TimeoutError()
+                result = await wait_for(task, remaining_time)
+                yield result
+                resume_feeding.set()
+        finally:
+            feeder_task.cancel()
+            await _consume_cancelled_future(feeder_task)
+
+            finalization_tasks = []
+            while submitted_tasks.qsize() > 0:
+                task = await submitted_tasks.get()
+                if task is not None:
+                    task.cancel()
+                    finalization_tasks.append(create_task(
+                        _consume_cancelled_future(task),
+                    ))
+            if finalization_tasks:
+                await gather(*finalization_tasks)
+
+    async def shutdown(self, wait=True, *, cancel_futures=False) -> None:
+        if self._shutdown:
+            return
+        self._shutdown = True
+
+        if cancel_futures:
+            finalization_tasks = []
+            while not self._input_queue.empty():
+                work_item = self._input_queue.get_nowait()
+                if work_item is not None:
+                    work_item.future.cancel()
+                    finalization_tasks.append(create_task(
+                        _consume_cancelled_future(work_item.future),
+                    ))
+            if finalization_tasks:
+                await gather(*finalization_tasks)
```

Comment:
If the code waits for all futures cancellations, there is no need to create a task per future.
`for` loop with waiting futures one by one could be enough; it doesn't wait the last future longer than the current code.